txnsync/bitmask.go

```diff
@@ -113,27 +113,33 @@ func (b *bitmask) trimBitmask(entries int) {
 }
 
 // iterate through the elements of bitmask without expanding it.
-// If more than maxIndex set bits are found, return error. For each set value, call callback.
-func (b *bitmask) iterate(entries int, maxIndex int, callback func(int, int) error) error {
+// call the func(entriesCount, setBitIndex) for every set bit
+// countOfEntries: is the size of the array that entryIndex is accessing: entryIndex < countOfEntries
+// countOfSetBits: is the size of the array that setBitCounter is accessing: setBitCounter < countOfSetBits
+func (b *bitmask) iterate(countOfEntries int, countOfSetBits int, callback func(int, int) error) error {
 	option := 0
 	if len(*b) > 0 {
 		option = int((*b)[0])
 	} else { // nothing to iterate
 		return nil
 	}
-	index := 0
+	setBitCounter := 0
 	switch option {
 	case 0:
 		for i, v := range (*b)[1:] {
 			for j := 0; j < 8 && v > 0; j++ {
 				if v&1 != 0 {
-					if index >= maxIndex {
+					if setBitCounter >= countOfSetBits {
 						return errDataMissing
 					}
-					if err := callback(8*i+j, index); err != nil {
+					entryIndex := 8*i + j
+					if entryIndex >= countOfEntries {
+						return errIndexNotFound
+					}
```

Comment:
while this one is correct, there is no reason to perform this on every iteration.
before going into the `for i, v := ...` loop, you can check the max value of i, figure out the biggest j and see if these are greater or equal to `countOfEntries`.